### PR TITLE
Breaking changes for MOI v0.10

### PR DESCRIPTION
**Pull request by metab0t**
Hope CI is happy this time.

### CODE REVIEW

**Comment by odow** [comment]
Yeah it should run now.

File: src/MOI/MOI_wrapper.jl [hunk 1]
```diff
@@ -555,22 +555,32 @@ end
...
 function MOI.get(model::Optimizer, ::MOI.TimeLimitSec)
-    return -MOI.get(model, MOI.RawParameter("MAXTIME"))
+    # MOI.attribute_value_type(MOI.TimeLimitSec()) = Union{Nothing, Float64}
+    return convert(Float64, -MOI.get(model, MOI.RawOptimizerAttribute("MAXTIME")))
 end
 
-MOI.Utilities.supports_default_copy_to(::Optimizer, ::Bool) = true
+MOI.supports_incremental_interface(::Optimizer) = true
 
 function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike; kwargs...)
```
Comment (odow) [inline src/MOI/MOI_wrapper.jl:569]:
```suggestion
function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike)
```

File: src/MOI/MOI_wrapper.jl [hunk 2]
```diff
@@ -555,22 +555,32 @@ end
...
-    return MOI.Utilities.automatic_copy_to(dest, src; kwargs...)
+    return MOI.Utilities.default_copy_to(dest, src; kwargs...)
```
Comment (odow) [inline src/MOI/MOI_wrapper.jl:570]:
```suggestion
    return MOI.Utilities.default_copy_to(dest, src)
```

File: test/MathOptInterface/MOI_Wrapper.jl [hunk 3]
```diff
@@ -1,297 +1,188 @@
...
+    model = MOI.Bridges.full_bridge_optimizer(optimizer, Float64)
+    MOI.set(model, MOI.Silent(), true)
+    MOI.Test.runtests(
+        model,
+        MOI.Test.Config(atol = 1e-3, rtol = 1e-3),
+        exclude = String[
+            "test_objective_set_via_modify",
+            "test_model_ListOfConstraintAttributesSet",
+            "test_solve_conflict_feasible",
+            "test_objective_get_ObjectiveFunction_ScalarAffineFunction",
```
Comment (odow) [inline test/MathOptInterface/MOI_Wrapper.jl:30]: These look like they are bugs that need fixing

File: test/MathOptInterface/MOI_Wrapper.jl [hunk 4]
```diff
@@ -1,297 +1,188 @@
...
+            "_SecondOrderCone_",
```
Comment (odow) [inline test/MathOptInterface/MOI_Wrapper.jl:31]: Why are all these excluded?

File: test/MathOptInterface/MOI_Wrapper.jl [hunk 5]
```diff
@@ -1,297 +1,188 @@
...
+
+    # Getting the results before the conflict refiner has been called must return an error.
+    @test MOI.get(model, MOI.ConflictStatus()) == MOI.COMPUTE_CONFLICT_NOT_CALLED
+    @test_throws ErrorException MOI.get(model, MOI.ConstraintConflictStatus(), c1)
+
+    # Once it's called, no problem.
+    MOI.compute_conflict!(model)
+    @test MOI.get(model, MOI.ConflictStatus()) == MOI.NO_CONFLICT_FOUND
+    @test MOI.get(model, MOI.ConstraintConflictStatus(), c1) == MOI.NOT_IN_CONFLICT
+    @test MOI.get(model, MOI.ConstraintConflictStatus(), c2) == MOI.NOT_IN_CONFLICT
```
Comment (odow) [inline test/MathOptInterface/MOI_Wrapper.jl:172]: You can remove all of these `conflict` functions. They're now in MOI.

File: test/MathOptInterface/MOI_Wrapper.jl [hunk 6]
```diff
@@ -1,297 +1,188 @@
...
 end
 
-@testset "test_farkas_dual_min" begin
+function test_farkas_dual_min()
```
Comment (odow) [inline test/MathOptInterface/MOI_Wrapper.jl:175]: You can remove all of these `test_farkas_` functions, they're now in MOI.

File: test/MathOptInterface/MOI_Wrapper.jl [hunk 7]
```diff
@@ -1,297 +1,188 @@
...
+            "test_solve_conflict_feasible",
+            "test_objective_get_ObjectiveFunction_ScalarAffineFunction",
+            "_SecondOrderCone_",
+            "test_constraint_PrimalStart_DualStart_SecondOrderCone",
+            "_RotatedSecondOrderCone_",
+            "_GeometricMeanCone_",
+            # It seems that Xpress cannot handle nonconvex quadratic constraint
+            "test_quadratic_nonconvex_",
+            # TODO: conflict tests should be investigated further
+            "test_solve_conflict_"
```
Comment (odow) [inline test/MathOptInterface/MOI_Wrapper.jl:38]: These need investigating

**Review by odow — APPROVED** [review]
Nice! Things are working at least. There are still a few tests being excluded, but it might be easiest to merge this PR first, and then tackle them individually.

**Comment by metab0t** [comment]
The callback tests now all pass.
I use `Revise.jl` and forget to restart `Julia` yesterday :(

**Comment by metab0t** [comment]
Conflict tests and farkas tests that exists in MOI.Test are removed.

File: test/MathOptInterface/MOI_Wrapper.jl (hunk 6 above)
Comment (metab0t) [reply test/MathOptInterface/MOI_Wrapper.jl]: Removed.

File: test/MathOptInterface/MOI_Wrapper.jl (hunk 5 above)
Comment (metab0t) [reply test/MathOptInterface/MOI_Wrapper.jl]: Removed.

**Comment by metab0t** [comment]
The MOI test config is similar with Gurobi.jl. The conic tests throw error if `MOI.ConstraintDual` and `MOI.DualObjectiveValue` are not excluded.

**Comment by metab0t** [comment]
I have a look at [MOI.Test.test_solve_conflict_EqualTo](https://github.com/jump-dev/MathOptInterface.jl/blob/2d76d9c8e1b2c6e77490dcc097acfca514b40315/src/Test/test_solve.jl#L1146-L1146).
The IIS obtained by Xpress says that b2 is not in conflict.

**Comment by joaquimg** [comment]
> I have a look at MOI.Test.test_solve_conflict_EqualTo.
> The IIS obtained by Xpress says that b2 is not in conflict.

Xpress IIS does not track variable bounds

File: test/MathOptInterface/MOI_Wrapper.jl (hunk 7 above)
Comment (joaquimg) [reply test/MathOptInterface/MOI_Wrapper.jl:38]: - Xpress does not accept nonconvex quadratic.
 
 - Xpress does not handle bounds properly in the IIS utility

File: test/MathOptInterface/MOI_Wrapper.jl (hunk 3 above)
Comment (joaquimg) [reply test/MathOptInterface/MOI_Wrapper.jl:30]: These should work indeed.

File: test/MathOptInterface/MOI_Wrapper.jl (hunk 4 above)
Comment (joaquimg) [reply test/MathOptInterface/MOI_Wrapper.jl:31]: These should work indeed.

**Comment by joaquimg** [comment]
@metab0t do you have plans to look into the failing tests?

**Comment by metab0t** [comment]
@joaquimg Yes, I plan to look at these problems.
I will look at these failing tests today.
```
"test_objective_set_via_modify"
"test_model_ListOfConstraintAttributesSet"
"test_objective_get_ObjectiveFunction_ScalarAffineFunction"
```

**Comment by metab0t** [comment]
@joaquimg
For `test_objective_set_via_modify`  the error happens [here](https://github.com/jump-dev/MathOptInterface.jl/blob/dcbc5275e9aac42030fda378f3cdfa0fc9d17026/src/Test/test_objective.jl#L517)
We need to change `model.objective_type` in
```
function MOI.modify(
    model::Optimizer,
    c::MOI.ConstraintIndex{MOI.ScalarAffineFunction{Float64}, <:Any},
    chg::MOI.ScalarCoefficientChange{Float64}
)
```
I cannot think of a good solution now.
Another error happens in line 514, the empty model should have zero attributes when calling `MOI.get(model, MOI.ListOfModelAttributesSet())` and I have fixed it.

**Comment by metab0t** [comment]
@joaquimg 
For `test_objective_get_ObjectiveFunction_ScalarAffineFunction`, there is a corner case in Xpress.jl.
If the `MOI.ObjectiveFunction` is set but `MOI.ObjectiveSense` is not set, then `model.is_feasibility=true` and `MOI.get(model::Optimizer, ::MOI.ObjectiveFunctionType) = nothing`.
Should we add another field to indicate whether objective function is set because `is_feasibility` has been used to indicate whether objective sense is set?

**Comment by odow** [comment]
Take a look at GLPK etc. 
https://github.com/jump-dev/GLPK.jl/blob/d11ae27aaa3666217786850c564aeffa138a6dac/src/MOI_wrapper/MOI_wrapper.jl#L88-L89

I switched to having a `objective_sense::Union{Nothing,MOI.OptimizationSense}`.

**Comment by metab0t** [comment]
> 
> 
> Take a look at GLPK etc. https://github.com/jump-dev/GLPK.jl/blob/d11ae27aaa3666217786850c564aeffa138a6dac/src/MOI_wrapper/MOI_wrapper.jl#L88-L89
> 
> I switched to having a `objective_sense::Union{Nothing,MOI.OptimizationSense}`.

I have made similar changes in latest commit.

**Comment by metab0t** [comment]
@joaquimg Any plans to merge this PR and prepare a version bump?

**Comment by joaquimg** [comment]
Still requires:
https://github.com/jump-dev/MathOptInterface.jl/pull/1648

**Comment by joaquimg** [comment]
Now waiting for:
https://github.com/jump-dev/MathOptInterface.jl/pull/1650

**Pull request merged**